src/ace-lib/ace.ml

```diff
@@ -77,15 +82,34 @@ let get_contents ?range e =
       let r = create_range (create_position r1 c1) (create_position r2 c2) in
       Js.to_string @@ document##(getTextRange r)
 
-let create_editor editor_div =
+let set_synchronized_status editor status =
+  List.iter (fun obs -> obs status) editor.sync_observers;
+  editor.synchronized <- status
+
+let create_editor editor_div check_valid_state =
   let editor = edit editor_div in
   Js.Unsafe.set editor "$blockScrolling" (Js.Unsafe.variable "Infinity");
   let data =
-    { editor; editor_div; marks = []; keybinding_menu = false; } in
+    { editor; editor_div;
+      marks = [];
+      keybinding_menu = false;
+      synchronized = true;
+      sync_observers = []
+    }
+  in
   editor##.customData := (data, None);
   editor##setOption (Js.string "displayIndentGuides") (Js.bool false);
+  editor##on (Js.string "change") (fun () ->
+      check_valid_state (set_contents data) ;
+      set_synchronized_status data false);
```

Comment:
BTW note that if we use the snippet suggested in [my other comment](https://github.com/ocaml-sf/learn-ocaml/pull/372/files#r710583507), then because of `lwt_alert`, we should refocus on the editor afterwards, for example here:

```suggestion
      set_synchronized_status data false;
      editor##focus);
```

Reply:
Done in 6b63b7d